### PR TITLE
Invoice Plugin Example

### PR DESCRIPTION
**Pull request by KBbitsP**
Added example to add tax item to invoice using InvoicePluginApi

@pierre @sbrossie @learnjavawithreshma

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+class HelloWorldInvoicePlugn extends PluginInvoicePluginApi  implements OSGIKillbillEventHandler {
+	
+	public HelloWorldInvoicePlugn(OSGIKillbillAPI killbillAPI, OSGIConfigPropertiesService configProperties, Clock clock) {
+		super(killbillAPI, configProperties, clock);
+		// TODO Auto-generated constructor stub
+	}
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:30]: Could you delete the `//TODO` statement?

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+		UUID accountId = newInvoice.getAccountId();
+        Account account = getAccount(accountId, callContext);
+		Set<Invoice> invoices =  allInvoicesOfAccount(account, newInvoice, callContext);
+		
+		ImmutableList.Builder<InvoiceItem> additionalItems  = ImmutableList.builder();
+		for(Invoice invoice : invoices) {
+			if (invoice.equals(newInvoice)) {
+                ImmutableList.Builder<InvoiceItem> newIntItems = ImmutableList.builder();
+                BigDecimal charge=new BigDecimal("100");
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:43]: Maybe rename `newIntItems` to something like `newInvoiceItems`?

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		// TODO Auto-generated constructor stub
+	}
+
+	@Override
+	public List<InvoiceItem> getAdditionalInvoiceItems(final Invoice newInvoice, final boolean dryRun, 
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+		UUID accountId = newInvoice.getAccountId();
+        Account account = getAccount(accountId, callContext);
+		Set<Invoice> invoices =  allInvoicesOfAccount(account, newInvoice, callContext);
+
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:38]: Is it necessary to retrieve account invoices (`allInvoicesOfAccount`) and loop through them?

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	public List<InvoiceItem> getAdditionalInvoiceItems(final Invoice newInvoice, final boolean dryRun, 
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+		UUID accountId = newInvoice.getAccountId();
+        Account account = getAccount(accountId, callContext);
+		Set<Invoice> invoices =  allInvoicesOfAccount(account, newInvoice, callContext);
+		
+		ImmutableList.Builder<InvoiceItem> additionalItems  = ImmutableList.builder();
+		for(Invoice invoice : invoices) {
+			if (invoice.equals(newInvoice)) {
+                ImmutableList.Builder<InvoiceItem> newIntItems = ImmutableList.builder();
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:42]: If at all it is necessary to retrieve account invoices and check for a match, I would suggest using the `invoiceId` (`invoice.getId.equals(newInvoice.getId()`)

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                BigDecimal charge=new BigDecimal("100");
+                //For generating tax item for each invoice item in new invoice
+                for (InvoiceItem item : newInvoice.getInvoiceItems()) {
+                	InvoiceItem newItem  = PluginInvoiceItem.createTaxItem(item, item.getInvoiceId(),invoice.getInvoiceDate(), null ,charge, "Tax Item");
+                	newIntItems.add(newItem);
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:47]: Maybe get rid of the `for` loop and add only one tax item per invoice to keep it simple.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                }
+                additionalItems.addAll(newIntItems.build());
+            }
+		}
+		return additionalItems.build();
+	}
+
+	private Set<Invoice> allInvoicesOfAccount(Account account, Invoice newInvoice, TenantContext tenantCtx) {
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:55]: Rename to `getAllAccountInvoices` or something similar that indicates what the method does.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        ImmutableSet.Builder<Invoice> builder = ImmutableSet.builder();
+        builder.addAll(getInvoicesByAccountId(account.getId(), tenantCtx));
+        builder.add(newInvoice);
+        return builder.build();
+    }
+
+	@Override
+	public void handleKillbillEvent(ExtBusEvent killbillEvent) {
+		// TODO Auto-generated method stub		
+	}
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:65]: Remove `//TODO` statement

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import org.killbill.billing.plugin.api.invoice.PluginInvoiceItem;
+import org.killbill.billing.plugin.api.invoice.PluginInvoicePluginApi;
+import org.killbill.billing.util.callcontext.CallContext;
+import org.killbill.billing.util.callcontext.TenantContext;
+import org.killbill.clock.Clock;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+class HelloWorldInvoicePlugn extends PluginInvoicePluginApi  implements OSGIKillbillEventHandler {
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:25]: Wondering if this class should be renamed to `HelloWorldInvoicePluginApi`?

File: pom.xml [hunk 9]
```diff
@@ -101,6 +101,11 @@
             <artifactId>killbill-platform-osgi-bundles-lib-killbill</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.kill-bill.billing.plugin</groupId>
+            <artifactId>killbill-plugin-api-invoice</artifactId>
+            <scope>provided</scope>
+        </dependency>
         <dependency>
```
Comment (reshmabidikar) [inline pom.xml:109]: Is this dependency intentionally deleted?

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java (hunk 4 above)
Comment (reshmabidikar) [reply src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:42]: Also, after a match is found, you can `break` out of the `for` loop as there will only be one invoice with the same id.

**Review by reshmabidikar — COMMENTED** [review]
Looks good @KBbitsP. I've left a few minor comments.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java (hunk 3 above)
Comment (KBbitsP) [reply src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePlugin.java:38]: Hi Reshma,

I used SimpleTax Plugin as a reference and i wanted to retain the same functionality of looping through invoices and creating additional items for new/historical invoices as required.

File: pom.xml (hunk 9 above)
Comment (KBbitsP) [reply pom.xml:109]: Hi Reshma,

It is needed as we need to use PluginInvoicePluginApi class for HelloWorldInvoicePluginApi

Comment (reshmabidikar) [reply pom.xml:109]: Ahh sorry, I thought you had deleted this dependency 😀

**Review by reshmabidikar — COMMENTED** [review]
Looks great, One final observation: The code does not seem to be formatted as per Kill Bill coding guidelines. Could you please format the code as explained [here](https://github.com/killbill/killbill-ee/wiki/Kill-Bill-Coding-Guidelines#code-formatting)?

**Comment by KBbitsP** [comment]
> Looks great, One final observation: The code does not seem to be formatted as per Kill Bill coding guidelines. Could you please format the code as explained [here](https://github.com/killbill/killbill-ee/wiki/Kill-Bill-Coding-Guidelines#code-formatting)?

Hi @reshmabidikar this link seems to throw error for me.

**Comment by KBbitsP** [comment]
Closing this one. Opened a new one.

**Comment by KBbitsP** [comment]
Hi @reshmabidikar added the example for Tax Item as well as Adjustment Item creation.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+		for (Invoice invoice : allInvoices) {
+			if (!invoice.getId().equals(newInvoice.getId())) {
+				List<InvoiceItem> invoiceItems = invoice.getInvoiceItems();
+				// Check for if any adjustment item exists for Historical Invoice
+				if (checkforAdjustmentItem(invoiceItems)) {
+					break;
+				}
+				for (InvoiceItem item : invoiceItems) {
+					if (isTaxItem(item)) {
+						charge = new BigDecimal("-40");
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:97]: Is there any particular reason for checking if this is a tax item?

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+				newInvoice.getInvoiceDate(), null, charge, "External Item", InvoiceItemType.EXTERNAL_CHARGE);
+		additionalItems.add(externalItem);
+
+		// Creating adjustment item for first Item of Historical Invoice
+		for (Invoice invoice : allInvoices) {
+			if (!invoice.getId().equals(newInvoice.getId())) {
+				List<InvoiceItem> invoiceItems = invoice.getInvoiceItems();
+				// Check for if any adjustment item exists for Historical Invoice
+				if (checkforAdjustmentItem(invoiceItems)) {
+					break;
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:93]: Is the `checkforAdjustmentItem` necessary? AFAIK, even if an adjustment item exists for an invoice item, you can still create another adjustment item.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	 *                   of the API request that has triggered this code to run.
+	 * @param callCtx    The context in which this code is running.
+	 * @return A new immutable list of new tax items, or adjustments on existing tax
+	 *         items.
+	 */
+	@Override
+	public List<InvoiceItem> getAdditionalInvoiceItems(final Invoice newInvoice, final boolean dryRun,
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+
+		UUID accountId = newInvoice.getAccountId();
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:69]: The code does not look to be formatted as per Kill Bill coding standards here and elsewhere.  As per Kill Bill coding standards, a variable is declared as `final`, whereas `accountId`  (and other variables in the code) are not declared as `final`.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+	 * @param invoiceItems
+	 * @return
+	 */
+	private boolean checkforAdjustmentItem(List<InvoiceItem> invoiceItems) {
+		boolean adjustmentItemPresent = false;
+		for (InvoiceItem invoiceItem : invoiceItems) {
+			if (invoiceItem.getInvoiceItemType().equals(InvoiceItemType.ITEM_ADJ)) {
+				adjustmentItemPresent = true;
+			}
+		}
```
Comment (reshmabidikar) [inline src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:139]: Isn't there a bug here? If an `ITEM_ADJ` is found, shouldn't we break out of the for loop?

Comment (KBbitsP) [reply src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:139]: Its not a bug. But yeah, you are right. We have put the break statement.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java (hunk 11 above)
Comment (KBbitsP) [reply src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:93]: Yes. We can still create the adjustment item. Our intention was to give an overview of how adjustment item is created. And, as mentioned in the comment, we are adding adjustment item for only first historical invoice.

File: src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java (hunk 10 above)
Comment (KBbitsP) [reply src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java:97]: Its not needed. We have removed this condition

**Review by reshmabidikar — APPROVED** [review]
Looks good!

**Pull request merged**